separating onnx parsing from burn-import

skewballfox opened the pull request:
## Pull Request Template

### Checklist
- [ ] confirm we want the name to be onnx-ir
- [x] deduplicate and remove unused dependencies
- [x] Confirmed that `run-checks all` script has been executed.
- [ ] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

#1812

### Changes

Separated out all the onnx specific code in burn-import to a new crate `onnx-ir` (which we can rename to whatever), and made some necessary changes to burn-import since `OnnxGraph` and other ir structs are now external. I also aliased a few types to just better distinguish between onnx and burn types.

some code had to be duplicated (`flatten_config`)or made public for the sake of using between both (`constant_conversion`)

### Testing

so far I've tested with onnx-test and burn-import. All test seem to be passing. examples seem okay

I think we should rework how onnx test are being used so that the models, and test themselves could be potentially reused. I think we can do that separately from this PR though.

Python side that would mean serializing graph inputs and outputs, burn side I'm guessing some macros for op types (such as `binary_op_test!`) though I'm open to suggestions if you guys want to handle it another way.

antimora commented on crates/onnx-ir/Cargo.toml:
```diff
@@ -0,0 +1,34 @@
+
+
+[dependencies]
+#burn = { path = "../burn", version = "0.14.0", features = ["ndarray"] }
```
Dead code. Also please check if all dependencies still needed.

skewballfox replied:
done

antimora commented on crates/onnx-ir/Cargo.toml:
```diff
@@ -0,0 +1,34 @@
+serde = { workspace = true, features = ["derive"] }
+strum = { workspace = true }
+strum_macros = { workspace = true }
+#tracing-core = { workspace = true }
+#tracing-subscriber = { workspace = true }
```
Dead

antimora reviewed (commented):
Looks good. A minor comment to review/clean up the dependencies. 

But I think it's almost ready to be approve. 

Since this is a brand new crate, we also need to update `publish.yml` under `.github.yml`. 

Also, can you please double check `adding-a-new-operation-to-burn.md` is up to date? Not sure if we have any references to about the onnx portion. 

@nathanielsimard @laggui can you sign off on the name?

laggui commented:
`onnx-ir` sounds good to me! Will try to take some time to review this.

skewballfox commented:
> Also, can you please double check `adding-a-new-operation-to-burn.md` is up to date? Not sure if we have any references to about the onnx portion.
 
I rewrote it, but I'm having trouble figuring out the best way to word it. I'd definitely appreciate a proof reader on the changed section. 

Also I added a readme though it's currently pretty barebones still figuring out what information it should have for other implementors.

laggui reviewed (commented):
Totally forgot about this PR, I'll hold off on merging an new ONNX ops if we're going to merge this soon otherwise you'll always have to fix the conflicts 😅 

Otherwise, the refactor LGTM.

antimora reviewed (approved):
Approving in advance. There are a couple minor text fixes but looks good overall.

skewballfox commented on contributor-book/src/guides/adding-a-new-operation-to-burn.md:
```diff
@@ -199,11 +199,9 @@ Generating the ONNX test files or tests is already covered
-defined under `src/burn/node`.
+Changes will need to be made to both `onnx-ir` and `burn-import`. The code within `onnx-ir` defines how to parse the nodes in an onnx file and produces the intermediate representation. The code within `burn-import` is divided into two sections: `src/onnx` and `src/burn`. The code under the former maps that intermediate representation to one used for code generation and the latter defines how to g ...
+
+So when you are loading a model, the operator is first parsed to an intermediate representation defined by `burn-import` and then mapped to a Burn operation defined under `src/burn/node`. that mapping is defined in `src/onnx/to_burn`
```
Formatted and changed the punctuation, wording

antimora commented:
Merged! Thanks for your work. I hope this crate will be useful to other project as well.
